File: app/src/main/java/com/babestudios/companyinfouk/ui/charges/ChargesPresenter.java
package com.babestudios.companyinfouk.ui.charges;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import android.util.Log;

import com.babestudios.companyinfouk.BuildConfig;
import com.babestudios.companyinfouk.data.DataManager;
import com.babestudios.companyinfouk.data.model.charges.Charges;

import net.grandcentrix.thirtyinch.TiPresenter;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import retrofit2.HttpException;

public class ChargesPresenter extends TiPresenter<ChargesActivityView> implements Observer<Charges> {

	DataManager dataManager;

	@Inject
	public ChargesPresenter(DataManager dataManager) {
		this.dataManager = dataManager;
	}

	@Override
	protected void onCreate() {
		super.onCreate();
	}

	@Override
	protected void onAttachView(@NonNull ChargesActivityView view) {
		super.onAttachView(view);
		view.showProgress();
		getCharges();
	}

	@VisibleForTesting
	public void getCharges() {
		dataManager.getCharges(getView().getCompanyNumber(), "0").subscribe(this);
	}

	public void loadMoreCharges(int page) {
		dataManager.getCharges(getView().getCompanyNumber(), String.valueOf(page * Integer.valueOf(BuildConfig.COMPANIES_HOUSE_SEARCH_ITEMS_PER_PAGE))).subscribe(this);
	}

	@Override
	public void onComplete() {
	}

	@Override
	public void onSubscribe(Disposable d) {

	}

	@Override
	public void onError(Throwable e) {
		getView().hideProgress();
		Log.d("test", "onError: " + e.fillInStackTrace());
		if (e instanceof HttpException) {
			HttpException h = (HttpException) e;
			if (h.code() == 404) {
				getView().showNoCharges();
			} else {
				getView().showError();
			}
		} else {
			getView().showError();
		}
	}

	@Override
	public void onNext(Charges charges) {
		getView().hideProgress();
		getView().showCharges(charges);
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
	}


}

File: app/src/main/java/com/babestudios/companyinfouk/ui/charges/ChargesPresenter.kt
package com.babestudios.companyinfouk.ui.charges

import android.annotation.SuppressLint
import com.babestudios.base.mvp.BasePresenter
import com.babestudios.base.mvp.Presenter
import com.babestudios.base.rxjava.SingleObserverWrapper
import com.babestudios.companyinfouk.BuildConfig
import com.babestudios.companyinfouk.data.CompaniesRepository
import com.babestudios.companyinfouk.data.model.charges.Charges
import com.babestudios.companyinfouk.ui.charges.list.AbstractChargesVisitable
import com.babestudios.companyinfouk.ui.charges.list.ChargesVisitable
import com.uber.autodispose.AutoDispose
import io.reactivex.CompletableSource
import javax.inject.Inject

interface ChargesPresenterContract : Presenter<ChargesState, ChargesViewModel> {
	fun fetchCharges(companyNumber: String)
	fun loadMoreCharges(page: Int)
}

@SuppressLint("CheckResult")
class ChargesPresenter
@Inject
constructor(var companiesRepository: CompaniesRepository) : BasePresenter<ChargesState, ChargesViewModel>(), ChargesPresenterContract {

	override fun setViewModel(viewModel: ChargesViewModel, lifeCycleCompletable: CompletableSource?) {
		this.viewModel = viewModel
		this.lifeCycleCompletable = lifeCycleCompletable
		viewModel.state.value?.chargeItems?.let {
			sendToViewModel {
				it.apply {
					this.isLoading = false
					this.contentChange = ContentChange.CHARGES_RECEIVED
				}
			}
		} ?: run {
			sendToViewModel {
				it.apply {
					this.isLoading = true
				}
			}
			viewModel.state.value?.companyNumber?.also {
				fetchCharges(it)
			}
		}
	}

	override fun fetchCharges(companyNumber: String) {
		companiesRepository.fetchCharges(companyNumber, "0")
				.`as`(AutoDispose.autoDisposable(lifeCycleCompletable))
				.subscribeWith(object : SingleObserverWrapper<Charges>(this) {
					override fun onSuccess(reply: Charges) {
						sendToViewModel {
							it.apply {
								this.isLoading = false
								this.contentChange = ContentChange.CHARGES_RECEIVED
								this.chargeItems = convertToVisitables(reply)
								this.totalCount = reply.totalCount?.toInt()
							}
						}
					}
				})
	}

	private fun convertToVisitables(reply: Charges): List<AbstractChargesVisitable> {
		return ArrayList(reply.items.map { item -> ChargesVisitable(item) })
	}

	override fun loadMoreCharges(page: Int) {
		if (viewModel.state.value?.chargeItems == null || viewModel.state.value?.chargeItems!!.size < viewModel.state.value?.totalCount!!) {
			companiesRepository.fetchCharges(viewModel.state.value?.companyNumber
					?: "", (page * Integer.valueOf(BuildConfig.COMPANIES_HOUSE_SEARCH_ITEMS_PER_PAGE)).toString())
					.subscribeWith(object : SingleObserverWrapper<Charges>(this) {
						override fun onSuccess(reply: Charges) {
							val newList = viewModel.state.value?.chargeItems?.toMutableList()
							newList?.addAll(convertToVisitables(reply))
							sendToViewModel {
								it.apply {
									this.isLoading = false
									this.contentChange = ContentChange.CHARGES_RECEIVED
									newList?.toList()?.let { list -> this.chargeItems = list }
								}
							}
						}
					})
		}
	}

}